File: src/class08/MST/Prim/Prim1.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;
import java.io.PrintWriter;
import java.io.OutputStreamWriter;
import java.io.IOException;

// 提交时修改类名为Main
public class Prim1 {
    public static int MAXN = 5001;
    public static int MAXM = 400001;
    public static int[] head = new int[MAXN], next = new int[MAXM], to = new int[MAXM], weight = new int[MAXM];
    public static int cnt;

    public static int n, m, nodeCnt;

    // 改写堆：需要手写，因为java不支持直接修改优先队列内部的元素
    // 另外，需要反向索引表来找点在堆中的位置。需要实现改写堆调整优先级上浮的接口。

    // 改写堆
    // [0]:点，[1]:src->该点的权值
    public static int[][] heap = new int[MAXN][2];
    public static int heapSize;
    // 反向索引表
    // where[v] = -1, 当前节点从未进入过堆
    // where[v] = -2, 当前节点已经弹出过堆，换言之，已经添加进入最小生成树中
    // where[v] = i， 当前节点在堆中位置，即实际中在heap数组的位置。
    public static int[] where = new int[MAXN];

    public static void build() {
        for (int i = 0; i <= n; i++) {
            head[i] = 0;
            where[i] = -1;
        }
        cnt = 1;
        nodeCnt = 0;
        heapSize = 0;
    }

    public static void addEdge(int u, int v, int w) {
        next[cnt] = head[u];
        to[cnt] = v;
        weight[cnt] = w;
        head[u] = cnt++;
    }

    public static void heapify(int i) {
        int l = (i << 1) + 1;
        while (l < heapSize) {
            int best = l + 1 < heapSize && heap[l + 1][1] < heap[l][1] ? l + 1 : l;
            best = heap[best][1] < heap[i][1] ? best : i;
            if (best == i) {
                return;
            }
            swap(best, i);
            i = best;
            l = (i << 1) + 1;
        }
    }

    public static int u, v, w;

    public static void swap(int i, int j) {
        int a = heap[i][0];
        int b = heap[j][0];
        where[a] = j;
        where[b] = i;
        int[] tmp = heap[i];
        heap[i] = heap[j];
        heap[j] = tmp;
    }

    public static void pop() {
        u = heap[0][0];
        w = heap[0][1];
        swap(0, --heapSize);
        heapify(0);
        where[u] = -2;
        nodeCnt++;
    }

    public static void heapInsert(int i) {
        while (i > 0 && heap[i][1] < heap[(i - 1) >> 1][1]) {
            swap(i, (i - 1) >> 1);
            i = (i - 1) >> 1;
        }
    }

    public static void addOrUpdateOrIgnore(int e) {
        v = to[e];
        w = weight[e];
        if (where[v] == -1) { // 从未进入过堆
            heap[heapSize][0] = v;
            heap[heapSize][1] = w;
            where[v] = heapSize++;
            heapInsert(where[v]);
        } else if (where[v] >= 0) { // 在堆中，尝试更新权值
            if (w < heap[where[v]][1]) {
                heap[where[v]][1] = w;
                heapInsert(where[v]);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StreamTokenizer in = new StreamTokenizer(br);
        PrintWriter out = new PrintWriter(new OutputStreamWriter(System.out));

        while (in.nextToken() != StreamTokenizer.TT_EOF) {
            n = (int) in.nval;
            in.nextToken();
            m = (int) in.nval;
            build();
            for (int i = 0; i < m; i++) {
                in.nextToken();
                u = (int) in.nval;
                in.nextToken();
                v = (int) in.nval;
                in.nextToken();
                w = (int) in.nval;
                addEdge(u, v, w);
                addEdge(v, u, w);
            }

            int ans = prim();
            out.println(nodeCnt == n ? ans : "orz");
        }

        out.flush();
        out.close();
        br.close();
    }

    public static int prim() {
        nodeCnt = 1;
        where[1] = -2; // 节点1加入最小生成树
        for (int ei = head[1]; ei > 0; ei = next[ei]) {
            addOrUpdateOrIgnore(ei);
        }
        int ans = 0;
        while (heapSize != 0) {
            pop();
            ans += w;
            for (int ei = head[u]; ei > 0; ei = next[ei]) {
                addOrUpdateOrIgnore(ei);
            }
        }
        return ans;
    }
}

File: src/class08/MST/Prim/Prim1.cpp
#include<bits/stdc++.h>
using namespace std;

//数据范围
#define MAXN 5001
#define MAXM 400001

//读入输入的变量
int n,m;

//带权图链式前向星的模板
int head[MAXN], ne[MAXM], to[MAXM],weight[MAXM];
int cnt;

//改写堆和反向索引表
//where[i]=-1,表示i位置的顶点未加入堆中
//where[i]=-2,表示i位置顶点已经出堆了。
//where[i]>=0,表示i位置顶点在堆上，位置是heap数组中的where[i]
int heap[MAXN][2], where[MAXN];
int heapSize, nodeCnt; //nodeCnt是处理节点的个数


//base 建图，初始化堆，初始化反向索引表
void build(){
    cnt = 1;
    heapSize = nodeCnt = 0;
    for(int i=1;i<=n;i++){
        head[i] = 0;
        where[i] = -1;
    }
}

void add_edge(int u,int v,int w){
    ne[cnt] = head[u];
    to[cnt] = v;
    weight[cnt] = w;
    head[u] = cnt++;
}
//交换堆中i，j位置； 反向索引表对应也要交换，
static void swap_heap(int i,int j){
    int a =heap[i][0];
    int b =heap[j][0];
    where[a] = j;
    where[b] = i;
    int tmp_to = heap[i][0], tmp_q = heap[i][1];
    heap[i][0] = heap[j][0], heap[i][1] = heap[j][1];
    heap[j][0] = tmp_to, heap[j][1] = tmp_q;
} 
static void heapify(int i){
    //左孩子
    int l = (i<<1)+1;
    while(l<heapSize){
        //左和右
        int best = l+1<heapSize&&heap[l+1][1]<heap[l][1]?l+1:l;
        //父与子
        best = heap[best][1]<heap[i][1]?best:i;
        //决出胜负发现父亲是最强的， 不用交换
        if(best==i){
            return ;
        }
        //交换并更新
        swap_heap(best,i);
        i = best;
        l = (i<<1)+1;
    }
}
int u,w;
static void pop(){
    //取出堆顶元素
    u = heap[0][0];
    w = heap[0][1];
    //原先堆顶挪至尾部， 堆整体大小--
    swap_heap(0, --heapSize);
    //下沉调堆
    heapify(0);
    //更新索引
    where[u] = -2;
    nodeCnt++;
}

static void heapInsert(int i){
    //这里仅仅只有上浮调整， 非常简洁的写法。
    while(heap[i][1]<heap[(i-1)>>1][1]){
        swap_heap(i,(i-1)>>1);
        i = (i-1)>>1;
    }
}

//要么更新，要么添加入堆，要么无视。
//传参是边ei
static void add_update_ignore(int ei){
    int v = to[ei];
    int w = weight[ei];
    if(where[v]==-2){
        return;
    }else if(where[v]==-1){
        //尾插+上浮
        heap[heapSize][0] = v;
        heap[heapSize][1] = w;
        where[v] = heapSize++;
        heapInsert(where[v]);
    }else if(where[v]>=0){
        //调整优先级
        //由于小根堆且优先级调低， 因此上浮调整。
        heap[where[v]][1] = min(heap[where[v]][1], w);
        heapInsert(where[v]);
    }
}
//反向索引表+手写堆 优化的Prim算法
//时间复杂度:$O(n+m)+O((m+n)*logn)$
static int prim(){
    nodeCnt = 1;
    where[1] = -2;
    for(int ei = head[1]; ei > 0 ; ei = ne[ei]){
        add_update_ignore(ei);
    }
    int ans = 0;
    //堆为空则结束。
    while(heapSize!=0){
        pop();
        ans += w;
        for(int ei = head[u];ei > 0;ei = ne[ei]){
            add_update_ignore(ei);
        }
    }
    return ans;
}
int main(void){
    ios::sync_with_stdio(false);
    std::cin.tie(nullptr);
    cin>>n>>m;
    build();
    int u,v,w;
    for(int i=0;i<m;i++){
        cin>>u>>v>>w;
        add_edge(u,v,w);
        add_edge(v,u,w);
    }
    int ans = prim();
    if(nodeCnt!=n){
        puts("orz");
    }else{
        printf("%d\n", ans);
    }
    return 0;
}